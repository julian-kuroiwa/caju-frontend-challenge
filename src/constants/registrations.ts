import { Registration, StatusType } from "~/types/registration";

export const initialValues = {
  employeeName: '',
  email: '',
  cpf: '',
  admissionDate: '',
  status: StatusType.REVIEW,
} as Registration;