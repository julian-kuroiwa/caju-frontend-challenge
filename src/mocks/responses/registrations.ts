export const registrationsMock = [
    {
      id: "3",
      admissionDate: "22/10/2023",
      email: "jose@caju.com.br",
      employeeName: "José Leão",
      status: "REVIEW",
      cpf: "78502270001"
    },
    {
      id: "1",
      admissionDate: "22/10/2023",
      email: "filipe@caju.com.br",
      employeeName: "Filipe Marins",
      status: "APPROVED",
      cpf: "39487627388"
    },
    {
      id: "f621",
      employeeName: "Teste teste",
      email: "teste@teste.com",
      cpf: "94009007087",
      admissionDate: "2023-10-10",
      status: "REPROVED"
    }
  ]