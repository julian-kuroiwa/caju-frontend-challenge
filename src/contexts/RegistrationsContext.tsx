import { ReactNode, createContext, useContext, useState } from 'react';
import { toast } from 'react-toastify';
import request from '~/api';
import SpinnerLoadingScreen from '~/components/Loading';
import { initialValues } from '~/constants/registrations';
import {
  NewRegistration,
  Registration,
  StatusType,
} from '~/types/registration';

interface RegistrationContextProps {
  children: ReactNode;
}

interface RegistrationContextType {
  registrations: Registration[];
  getRegistrations: (props: string) => void;
  addNewRegistration: (props: NewRegistration) => void;
  handleStatus: () => void;
  handleRemove: () => void;
  loading: boolean;
  loadContent: () => void;
  setRegistrationNewStatus: (value: keyof typeof StatusType) => void;
  setCurrentRegistration: (value: Registration) => void;
  currentRegistration: Registration;
  registrationNewStatus: keyof typeof StatusType;
}

export const RegistrationContext = createContext({} as RegistrationContextType);
export const useRegistrationContext = () => useContext(RegistrationContext);

const RegistrationProvider = ({ children }: RegistrationContextProps) => {
  const [registrations, setRegistrations] = useState<Registration[]>([]);
  const [loading, setLoading] = useState(false);
  const [_, setRefetch] = useState(false);
  const [registrationNewStatus, setRegistrationNewStatus] = useState<
    keyof typeof StatusType
  >(StatusType.REVIEW);
  const [currentRegistration, setCurrentRegistration] =
    useState<Registration>(initialValues);

  const delayTurnOffLoading = () => {
    setTimeout(() => {
      setLoading(false);
    }, 300);
  };

  const addNewRegistration = async (newRegistration: NewRegistration) => {
    setLoading(true);

    try {
      await request.post('/registrations', newRegistration);
      toast('Cadastro realizado com sucesso!', { type: 'success' });
    } catch (err) {
      console.error(err);
      toast('Erro ao tentar criar um novo cadastro. Tente novamente.', {
        type: 'error',
      });
    } finally {
      delayTurnOffLoading();
    }
  };

  const getRegistrations = async (filter: string) => {
    setRefetch(false);
    setLoading(true);

    try {
      const { data } = await request.get(`/registrations?${filter}`);

      setRegistrations(data);
      if (filter) {
        toast('Filtro realizado com sucesso!', { type: 'success' });

        return;
      }

      toast('Registros carregados com sucesso!', { type: 'success' });
    } catch (err) {
      console.error(err);
      toast('Erro ao tentar filtrar por CPF. Tente novamente.', {
        type: 'error',
      });
    } finally {
      delayTurnOffLoading();
    }
  };

  const handleRemove = async () => {
    setLoading(true);

    try {
      await request.delete(`/registrations/${currentRegistration?.id}`);

      const registrationsUpdated = registrations.filter(
        (registrationFiltered) =>
          currentRegistration?.id !== registrationFiltered.id,
      );

      setRegistrations(registrationsUpdated);
      toast(
        `Registro de ${currentRegistration.employeeName} removido com sucesso!`,
        { type: 'success' },
      );
    } catch (err) {
      console.error(err);
      toast(
        `Erro ao tentar excluir o registro de ${currentRegistration.employeeName}. Tente novamente mais tarde.`,
        {
          type: 'error',
        },
      );
    } finally {
      delayTurnOffLoading();
    }
  };

  const handleStatus = async () => {
    setLoading(true);

    try {
      const { data: registrationUpdated } = await request.put(
        `/registrations/${currentRegistration?.id}`,
        {
          ...currentRegistration,
          status: registrationNewStatus,
        },
      );

      const registrationsUpdated = registrations.map((registrationMapped) =>
        registrationMapped.id === currentRegistration?.id
          ? registrationUpdated
          : registrationMapped,
      );

      setRegistrations(registrationsUpdated);
      toast(
        `Status alterado de ${currentRegistration.employeeName} com sucesso!`,
        { type: 'success' },
      );
    } catch (err) {
      console.error(err);
      toast(
        `Erro ao tentar alterar o status de ${currentRegistration.employeeName} . Tente novamente mais tarde.`,
        {
          type: 'error',
        },
      );
    } finally {
      delayTurnOffLoading();
    }
  };

  return (
    <RegistrationContext.Provider
      value={{
        getRegistrations,
        registrations,
        handleStatus,
        handleRemove,
        loading,
        loadContent: () => setRefetch(true),
        addNewRegistration,
        setRegistrationNewStatus,
        registrationNewStatus,
        setCurrentRegistration,
        currentRegistration,
      }}>
      {children}
      <SpinnerLoadingScreen loading={loading} />
    </RegistrationContext.Provider>
  );
};

export { RegistrationProvider };
