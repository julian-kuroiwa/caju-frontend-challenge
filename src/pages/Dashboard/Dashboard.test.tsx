import {
  fireEvent,
  render,
  screen,
  waitFor,
  within,
} from '@testing-library/react';
import { RegistrationProvider } from '~/contexts/RegistrationsContext';
import { registrationsMock } from '~/mocks/responses/registrations';
import DashboardPage from '.';

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

jest.useFakeTimers();

const renderComponent = () =>
  render(
    <RegistrationProvider>
      <DashboardPage />
    </RegistrationProvider>,
  );

describe('Dashboard', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', async () => {
    const { container } = renderComponent();

    await screen.findAllByTestId('registration-card', { exact: false });

    expect(container).toMatchSnapshot();
  });

  it('should show cards on dashboard with registrations', async () => {
    renderComponent();

    const registrationCards = await screen.findAllByTestId(
      'registration-card',
      { exact: false },
    );

    expect(registrationCards.length).toEqual(registrationsMock.length);
  });

  it('should be able to filter the card by cpf', async () => {
    renderComponent();

    const value = '39487627388';
    const filteredRegistration = registrationsMock.filter(
      (registration) => registration.cpf === value,
    );
    const searchInput = screen.getByPlaceholderText('Digite um CPF vÃ¡lido');
    fireEvent.change(searchInput, { target: { value } });

    jest.advanceTimersByTime(1000);

    const registrationCards = await screen.findAllByTestId(
      'registration-card',
      { exact: false },
    );

    expect(registrationCards.length).toEqual(filteredRegistration.length);
    expect(
      screen.getByText(filteredRegistration[0]?.employeeName),
    ).toBeInTheDocument();
  });

  it('should be able to move to collumn "Reprovado" if click in "Reprovar" and confirm the action', async () => {
    renderComponent();

    const reproveButton = await screen.findByTestId(
      'change-status-modal-reproved-button',
    );

    const registrationsReproved = screen.getAllByTestId(
      'registration-card-REPROVED',
    );

    expect(registrationsReproved.length).toEqual(1);

    fireEvent.click(reproveButton);

    const confirmChangeStatus = await screen.findByRole('button', {
      name: 'Confirmar',
    });

    fireEvent.click(confirmChangeStatus);

    await waitFor(() => {
      const newRegistrationsReproved = screen.getAllByTestId(
        'registration-card-REPROVED',
      );
      expect(newRegistrationsReproved.length).toEqual(2);
    });
  });

  it('should be able to move to collumn "Aprovado" if click in "Aprovar" and confirm the action', async () => {
    renderComponent();

    const approveButton = await screen.findByTestId(
      'change-status-modal-approved-button',
    );

    const registrationsApproved = screen.getAllByTestId(
      'registration-card-APPROVED',
    );

    expect(registrationsApproved.length).toEqual(1);

    fireEvent.click(approveButton);

    const confirmChangeStatus = await screen.findByRole('button', {
      name: 'Confirmar',
    });

    fireEvent.click(confirmChangeStatus);

    await waitFor(() => {
      const newRegistrationsApproved = screen.getAllByTestId(
        'registration-card-APPROVED',
      );
      expect(newRegistrationsApproved.length).toEqual(2);
    });
  });

  it('should be able to move to collumn "Pronto para revisar" if click in "Revisar novamente" and confirm the action', async () => {
    renderComponent();

    const reviewButton = await screen.findAllByTestId(
      'change-status-modal-review-button',
    );

    const registrationsReview = screen.getAllByTestId(
      'registration-card-REVIEW',
    );

    expect(registrationsReview.length).toEqual(1);

    fireEvent.click(reviewButton[0]);

    const confirmChangeStatus = await screen.findByRole('button', {
      name: 'Confirmar',
    });

    fireEvent.click(confirmChangeStatus);

    await waitFor(() => {
      const newRegistrationsReview = screen.getAllByTestId(
        'registration-card-REVIEW',
      );
      expect(newRegistrationsReview.length).toEqual(2);
    });
  });

  it('should be able to remove a registration if trash can button is clicked and confirm the action', async () => {
    renderComponent();

    const registrationCards = await screen.findAllByTestId(
      'registration-card',
      { exact: false },
    );

    expect(registrationCards.length).toEqual(registrationsMock.length);

    const deleteButton = within(registrationCards[0]).getByTestId(
      'delete-button',
    );

    fireEvent.click(deleteButton);

    const removeConfirmationButton = await screen.findByRole('button', {
      name: 'Confirmar',
    });

    fireEvent.click(removeConfirmationButton);

    await waitFor(() => {
      const registrations = screen.getAllByTestId('registration-card', {
        exact: false,
      });
      expect(registrations.length).toEqual(2);
    });
  });
});
