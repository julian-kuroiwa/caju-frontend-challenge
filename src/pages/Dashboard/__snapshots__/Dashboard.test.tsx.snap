// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should match snapshot 1`] = `
<div>
  <div
    class="sc-fUnNpA gemtlB"
  >
    <div
      class="sc-cwHqhk fYffUJ"
    >
      <div>
        <label />
        <input
          class="sc-jlZhRR jsGjNp"
          placeholder="Digite um CPF válido"
          value=""
        />
        <span
          data-testid="error-message-undefined"
          style="font-size: 12px; color: red;"
        />
      </div>
      <div
        class="sc-jEABJj kgEuvB"
      >
        <button
          aria-label="refetch"
          class="sc-dAlxHm eFbcSi"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 20 20"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M4 2a1 1 0 011 1v2.101a7.002 7.002 0 0111.601 2.566 1 1 0 11-1.885.666A5.002 5.002 0 005.999 7H9a1 1 0 010 2H4a1 1 0 01-1-1V3a1 1 0 011-1zm.008 9.057a1 1 0 011.276.61A5.002 5.002 0 0014.001 13H11a1 1 0 110-2h5a1 1 0 011 1v5a1 1 0 11-2 0v-2.101a7.002 7.002 0 01-11.601-2.566 1 1 0 01.61-1.276z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <button
          class="sc-eqUzNf gvJhSY"
        >
          Nova Admissão
        </button>
      </div>
    </div>
    <div
      class="sc-imWZod iahWDi"
    >
      <div
        class="sc-jXbVAB iTQlrL"
        data-testid="REVIEW"
        status="REVIEW"
      >
        <h3
          class="sc-dhKdPU fzUDCd"
          status="REVIEW"
        >
          Pronto para revisar
        </h3>
        <div
          class="sc-kpDprT gEiEmr"
        >
          <div
            class="sc-dcJtft cMKrRQ"
            data-testid="registration-card-REVIEW"
          >
            <div
              class="sc-iGgVNO fbVpXc"
            >
              <svg
                aria-hidden="true"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16 7a4 4 0 11-8 0 4 4 0 018 0zM12 14a7 7 0 00-7 7h14a7 7 0 00-7-7z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
              <h3>
                José Leão
              </h3>
            </div>
            <div
              class="sc-iGgVNO fbVpXc"
            >
              <svg
                aria-hidden="true"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3 8l7.89 5.26a2 2 0 002.22 0L21 8M5 19h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v10a2 2 0 002 2z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
              <p>
                jose@caju.com.br
              </p>
            </div>
            <div
              class="sc-iGgVNO fbVpXc"
            >
              <svg
                aria-hidden="true"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 7V3m8 4V3m-9 8h10M5 21h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v12a2 2 0 002 2z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
              <span>
                22/10/2023
              </span>
            </div>
            <div
              class="sc-gsFSjX LLUum"
            >
              <button
                bgcolor="rgb(255, 145, 154)"
                class="sc-fqkwJk eGPNoB"
                data-testid="change-status-modal-reproved-button"
              >
                Reprovar
              </button>
              <button
                bgcolor="rgb(155, 229, 155)"
                class="sc-fqkwJk jCekve"
                data-testid="change-status-modal-approved-button"
              >
                Aprovar
              </button>
              <svg
                aria-hidden="true"
                class="sc-kAycRU gOPfrk"
                data-testid="delete-button"
                fill="none"
                height="1em"
                role="button"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-jXbVAB cVhtxp"
        data-testid="APPROVED"
        status="APPROVED"
      >
        <h3
          class="sc-dhKdPU fwWCsB"
          status="APPROVED"
        >
          Aprovado
        </h3>
        <div
          class="sc-kpDprT gEiEmr"
        >
          <div
            class="sc-dcJtft cMKrRQ"
            data-testid="registration-card-APPROVED"
          >
            <div
              class="sc-iGgVNO fbVpXc"
            >
              <svg
                aria-hidden="true"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16 7a4 4 0 11-8 0 4 4 0 018 0zM12 14a7 7 0 00-7 7h14a7 7 0 00-7-7z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
              <h3>
                Filipe Marins
              </h3>
            </div>
            <div
              class="sc-iGgVNO fbVpXc"
            >
              <svg
                aria-hidden="true"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3 8l7.89 5.26a2 2 0 002.22 0L21 8M5 19h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v10a2 2 0 002 2z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
              <p>
                filipe@caju.com.br
              </p>
            </div>
            <div
              class="sc-iGgVNO fbVpXc"
            >
              <svg
                aria-hidden="true"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 7V3m8 4V3m-9 8h10M5 21h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v12a2 2 0 002 2z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
              <span>
                22/10/2023
              </span>
            </div>
            <div
              class="sc-gsFSjX LLUum"
            >
              <button
                bgcolor="#ff8858"
                class="sc-fqkwJk cjmksH"
                data-testid="change-status-modal-review-button"
              >
                Revisar novamente
              </button>
              <svg
                aria-hidden="true"
                class="sc-kAycRU gOPfrk"
                data-testid="delete-button"
                fill="none"
                height="1em"
                role="button"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-jXbVAB cnqFim"
        data-testid="REPROVED"
        status="REPROVED"
      >
        <h3
          class="sc-dhKdPU fIZMKx"
          status="REPROVED"
        >
          Reprovado
        </h3>
        <div
          class="sc-kpDprT gEiEmr"
        >
          <div
            class="sc-dcJtft cMKrRQ"
            data-testid="registration-card-REPROVED"
          >
            <div
              class="sc-iGgVNO fbVpXc"
            >
              <svg
                aria-hidden="true"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16 7a4 4 0 11-8 0 4 4 0 018 0zM12 14a7 7 0 00-7 7h14a7 7 0 00-7-7z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
              <h3>
                Teste teste
              </h3>
            </div>
            <div
              class="sc-iGgVNO fbVpXc"
            >
              <svg
                aria-hidden="true"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3 8l7.89 5.26a2 2 0 002.22 0L21 8M5 19h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v10a2 2 0 002 2z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
              <p>
                teste@teste.com
              </p>
            </div>
            <div
              class="sc-iGgVNO fbVpXc"
            >
              <svg
                aria-hidden="true"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 7V3m8 4V3m-9 8h10M5 21h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v12a2 2 0 002 2z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
              <span>
                09/10/2023
              </span>
            </div>
            <div
              class="sc-gsFSjX LLUum"
            >
              <button
                bgcolor="#ff8858"
                class="sc-fqkwJk cjmksH"
                data-testid="change-status-modal-review-button"
              >
                Revisar novamente
              </button>
              <svg
                aria-hidden="true"
                class="sc-kAycRU gOPfrk"
                data-testid="delete-button"
                fill="none"
                height="1em"
                role="button"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="sc-aYaIB jZrxtb"
    date-testid="loader"
  >
    <div
      class="sc-gEvDqW gEDMht"
    />
  </div>
</div>
`;
